Source/WTF/wtf/glib/RunLoopGLib.cpp

```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
 RunLoop::~RunLoop()
 {
     g_source_destroy(m_source.get());
+    m_shouldStop = true;
+}
+
+void RunLoop::runGLibMainLoop(GMainContext* mainContext)
+{
+    g_main_context_ref(mainContext);
+    g_main_context_push_thread_default(mainContext);
+    notifyActivity(Activity::Entry);
+
+    Vector<GPollFD, 32> pollFDs;
+    while (!m_shouldStop) {
+        gint maxPriority = 0;
+        g_main_context_prepare(mainContext, &maxPriority);
+
+        auto numFDs = g_main_context_query(mainContext, maxPriority, nullptr, nullptr, 0);
+        pollFDs.clear();
+        pollFDs.resize(numFDs);
+        auto* pollFDsData = pollFDs.mutableSpan().data();
+
+        gint timeout = 0;
+        g_main_context_query(mainContext, maxPriority, &timeout, pollFDsData, numFDs);
 
-    for (int i = m_mainLoops.size() - 1; i >= 0; --i) {
-        if (!g_main_loop_is_running(m_mainLoops[i].get()))
-            continue;
-        g_main_loop_quit(m_mainLoops[i].get());
+        notifyActivity(Activity::BeforeWaiting);
+
+        auto result = g_poll(pollFDsData, numFDs, timeout);
+        if (result < 0) {
+            LOG_ERROR("RunLoop::runGLibMainLoop() - polling failed, exiting.");
+            break;
+        }
+
+        notifyActivity(Activity::AfterWaiting);
+
+        g_main_context_check(mainContext, maxPriority, pollFDsData, numFDs);
+        g_main_context_dispatch(mainContext);
     }
+
+    notifyActivity(Activity::Exit);
+    g_main_context_unref(mainContext);
+    g_main_context_pop_thread_default(mainContext);
```

Comment:
This is the wrong order. If it's OK for the unref to not be last, then the unref is not needed. :) (And it's needed.) Another benefit of GRefPtr.